python-checks/src/main/java/org/sonar/python/checks/FunctionUsingLoopVariableCheck.java

```diff
@@ -70,31 +73,54 @@ private static void checkFunctionLike(SubscriptionContext ctx) {
         Tree usageTree = usage.tree();
         if (isUsedInFunctionLike(usageTree, functionLike) && !usage.isBindingUsage()) {
+          if (TreeUtils.firstAncestor(usageTree, t -> t.is(Tree.Kind.NONLOCAL_STMT, Tree.Kind.GLOBAL_STMT)) != null) {
+            // We don't raise any issue on the variable if it's part of a nonlocal or global statement
+            problematicUsages.clear();
+            break;
+          }
           problematicUsages.add(usageTree);
         }
         if (usage.isBindingUsage() && isWithinEnclosingLoop(usageTree, enclosingLoop)) {
           bindingUsages.add(usageTree);
         }
       }
-      reportIssue(ctx, functionLike, problematicUsages, bindingUsages);
+      reportIssue(ctx, functionLike, problematicUsages, bindingUsages, symbol.name());
     }
   }
 
-  private static void reportIssue(SubscriptionContext ctx, FunctionLike functionLike, List<Tree> problematicUsages, List<Tree> bindingUsages) {
+  private static void reportIssue(SubscriptionContext ctx, FunctionLike functionLike, List<Tree> problematicUsages, List<Tree> bindingUsages, String symbolName) {
     if (!problematicUsages.isEmpty() && !bindingUsages.isEmpty()) {
-      PreciseIssue issue = ctx.addIssue(problematicUsages.get(0), "Pass this variable as a parameter with a default value.")
-        .secondary(bindingUsages.get(bindingUsages.size() - 1), "Assignment in the loop");
+      PreciseIssue issue;
       if (functionLike.is(Tree.Kind.FUNCDEF)) {
-        issue.secondary(((FunctionDef) functionLike).name(), "Function definition");
+        issue = ctx.addIssue(problematicUsages.get(0), String.format("Add a parameter to function \"%s\" and use variable \"%s\" as its default value;" +
+          "The value of \"%s\" might change at the next loop iteration.", ((FunctionDef) functionLike).name().name(), symbolName, symbolName))
+          .secondary(((FunctionDef) functionLike).name(), "Function capturing the variable");
+      } else {
+        issue = ctx.addIssue(problematicUsages.get(0),
+          String.format("Add a parameter to the parent lambda function and use variable \"%s\" as its default value; " +
+            "The value of \"%s\" might change at the next loop iteration.", symbolName, symbolName))
+          .secondary(((LambdaExpression) functionLike).lambdaKeyword(), "Lambda capturing the variable");
+      }
+      for (Tree bindingUsage : bindingUsages) {
+        issue.secondary(bindingUsage, "Assignment in the loop");
       }
     }
   }
 
   private static boolean isUsedInFunctionLike(Tree usageTree, FunctionLike functionLike) {
     ParameterList parameters = functionLike.parameters();
-    if (parameters != null && parameters.nonTuple().stream().anyMatch(p -> usageTree.equals(p.defaultValue()))) {
+    if (parameters != null && parameters.nonTuple().stream().anyMatch(p ->
+      p.defaultValue() != null && (usageTree.equals(p.defaultValue()) || TreeUtils.hasDescendant(p.defaultValue(), t -> t.equals(usageTree))))) {
```

Comment:
I think it's better to extract this lambda expression in a separated method

Reply:
Done in https://github.com/SonarSource/sonar-python/pull/667/commits/cecb37ed1c5dd391526a50dc274519acffd623a4